File: src/main/java/io/bindernews/thegrackle/power/FireheartPower.java
package io.bindernews.thegrackle.power;

import com.megacrit.cardcrawl.actions.common.RemoveSpecificPowerAction;
import com.megacrit.cardcrawl.core.AbstractCreature;
import com.megacrit.cardcrawl.powers.AbstractPower;
import io.bindernews.thegrackle.GrackleMod;
import io.bindernews.thegrackle.stance.StancePhoenix;

import static io.bindernews.thegrackle.helper.ModInterop.iop;
import static io.bindernews.thegrackle.patches.Fields.fireheartGained;

public class FireheartPower extends BasePower {
    public static final String POWER_ID = GrackleMod.makeId(FireheartPower.class);
    public static int fireheartRequired = 10;

    public FireheartPower(AbstractCreature owner, int amount) {
        super(POWER_ID);
        setOwnerAmount(owner, amount);
        isTurnBased = true;
        type = PowerType.BUFF;
        fireheartGained.set(owner, fireheartGained.get(owner) + amount);
        loadRegion("flameBarrier");
        updateDescription();
    }

    @Override
    public void stackPower(int stackAmount) {
        super.stackPower(stackAmount);
        // Yes, this means you can build fireheart even while in stance.
        if (amount >= fireheartRequired && canEnterPhoenixStance()) {
            addToBot(new RemoveSpecificPowerAction(owner, owner, this));
            addToBot(iop().changeStance(owner, StancePhoenix.STANCE_ID));
        }
    }

    public boolean canEnterPhoenixStance() {
        return !StancePhoenix.is(iop().getStance(owner)) && !owner.hasPower(CoolingPhoenixPower.POWER_ID);
    }

    @Override
    public AbstractPower makeCopy() {
        return new FireheartPower(owner, amount);
    }
}

File: src/main/kotlin/io/bindernews/thegrackle/power/FireheartPower.kt
package io.bindernews.thegrackle.power

import com.megacrit.cardcrawl.actions.common.RemoveSpecificPowerAction
import com.megacrit.cardcrawl.core.AbstractCreature
import com.megacrit.cardcrawl.powers.AbstractPower
import io.bindernews.thegrackle.GrackleMod
import io.bindernews.thegrackle.helper.ModInterop.Companion.iop
import io.bindernews.thegrackle.helper.fireheartGained
import io.bindernews.thegrackle.stance.StancePhoenix
import io.bindernews.thegrackle.stance.StancePhoenix.Companion.isStance

class FireheartPower(owner: AbstractCreature, amount: Int) : BasePower(POWER_ID) {
    init {
        setOwnerAmount(owner, amount)
        isTurnBased = true
        type = PowerType.BUFF
        owner.fireheartGained += amount
        loadRegion("flameBarrier")
        updateDescription()
    }

    override fun stackPower(stackAmount: Int) {
        super.stackPower(stackAmount)
        // Yes, this means you can build fireheart even while in stance.
        if (amount >= fireheartRequired && canEnterPhoenixStance()) {
            addToBot(RemoveSpecificPowerAction(owner, owner, this))
            addToBot(iop().changeStance(owner, StancePhoenix.STANCE_ID))
        }
    }

    fun canEnterPhoenixStance(): Boolean {
        return !isStance(iop().getStance(owner)) && !owner.hasPower(CoolingPhoenixPower.POWER_ID)
    }

    override fun makeCopy(): AbstractPower {
        return FireheartPower(owner, amount)
    }

    companion object {
        @JvmField val POWER_ID = GrackleMod.makeId(FireheartPower::class.java)
        @JvmField var fireheartRequired = 10
    }
}